task_manager: module: make_task: enter gate when the task is created

bhalevy opened the pull request:
Passing the gate_closed_exception to the task promise
ends up with abandoned exception since no-one is waiting
for it.

Instead, enter the gate when the task is made
so it will fail make_task if the gate is already closed.

Fixes scylladb/scylladb#15211

In addition, this series adds a private abort_source for each task_manager module
(chained to the main task_manager::abort_source) and abort is requested on task_manager::module::stop().

gate holding in compaction_manager is hardened
and makes sure to stop compaction_manager and task_manager in sstable_compaction_test cases.

Deexie commented on tasks/task_manager.cc:
```diff
@@ -180,17 +181,18 @@ void task_manager::task::start() {
     try {
         // Background fiber does not capture task ptr, so the task can be unregistered and destroyed independently in the foreground.
         // After the ttl expires, the task id will be used to unregister the task if that didn't happen in any other way.
-        (void)with_gate(_impl->_module->async_gate(), [f = done(), module = _impl->_module, id = id()] () mutable {
```
I think that `gate_closed_exception` is never really passed to `finish_failed` (it would do be handled if so).
It rather creates a future that would run in the background here, finishes with the `start` method and then at some point reactor gets back to this and throws but nothing handles it.

So, I suggest moving 
```
auto gate_holder = module->async_gate().hold();
```
under `try` and allow the task to record that the gate was closed

bhalevy replied:
I tried it and it didn't work.
Note that with_gate throws gate_closed_exception is the gate is closed.

Deexie replied:
The problem that appears when `hold()` is called outside of `try` is that a task will be created, registered but never finished and unregistered from task manager.

But this can happen on shutdown only anyway, so I guess it's ok

bhalevy replied:
moved gate holding to make_task time in v3

Deexie commented on compaction/compaction_manager.cc:
```diff
@@ -1201,13 +1215,14 @@ class regular_compaction_task_executor : public compaction_task_executor, public
 }
 
 void compaction_manager::submit(table_state& t) {
-    if (_state != state::enabled || t.is_auto_compaction_disabled_by_user()) {
```
How about `t.is_auto_compaction_disabled_by_user()`?

bhalevy replied:
fixed in v2 ([30ad31e](https://github.com/scylladb/scylladb/commit/30ad31ef01855f55ac5308829d49d5ea0dc560c7))

bhalevy commented:
In v3 ([7aa448d](https://github.com/scylladb/scylladb/commit/7aa448d23c86bfcac5fa450529eb5d65af652480)):
- task_manaer: module: stop: request abort
- task_manager: module: make_task: enter gate when the task is created

bhalevy commented on tasks/task_manager.cc:
```diff
@@ -8,6 +8,8 @@
 
```
It should have neen this way from the beginning...

bhalevy commented:
In v4 ([91397c4](https://github.com/scylladb/scylladb/commit/91397c44f152342e04e26e6ea20be54b3686414f)
- rebased
- task_manager: task::impl: subscribe to module about_source

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3362/

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3377/

bhalevy commented:
In v5 ([1262202](https://github.com/scylladb/scylladb/commit/12622028904ad683f521ff61e84c8a75f2383a9d)):
- rebased
- fixed double close of compaction_state gate

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3439/

bhalevy commented:
> In v5 ([1262202](https://github.com/scylladb/scylladb/commit/12622028904ad683f521ff61e84c8a75f2383a9d)):
> 
> * rebased
> * fixed double close of compaction_state gate

Apparently I failed to amend the last patch.
Here it is again: [cfecb68](https://github.com/scylladb/scylladb/commit/cfecb68245a9e97c3bd8b1cafd003d4a2dcbe49b)

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3458/

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3461/

bhalevy commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3461/

Failed on unrelated issue https://github.com/scylladb/scylladb/issues/15285

scylladb-promoter commented:
CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3479/

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3495/

bhalevy commented:
@scylladb/scylla-maint please consider merging
